pkg/capacityscheduling/capacity_scheduling.go

```diff
@@ -0,0 +1,882 @@
+		name := node.Node().Name
+		// We reply on the status by each plugin - 'Unschedulable' or 'UnschedulableAndUnresolvable'
+		// to determine whether preemption may help or not on the node.
+		if m[name].Code() == framework.UnschedulableAndUnresolvable {
+			continue
+		}
+		potentialNodes = append(potentialNodes, node)
+	}
+	return potentialNodes
+}
+
+func selectVictimsOnNode(
+	ctx context.Context,
+	ph framework.PreemptHandle,
+	state *framework.CycleState,
+	pod *v1.Pod,
+	nodeInfo *framework.NodeInfo,
+	pdbs []*policy.PodDisruptionBudget,
+	snapshotState *SnapshotState,
+	preFilterState *PreFilterState,
+) ([]*v1.Pod, int, bool) {
+	var potentialVictims []*v1.Pod
+	var moreThanMinWithPreemptor bool
+
+	removePod := func(rp *v1.Pod) error {
+		if err := nodeInfo.RemovePod(rp); err != nil {
+			return err
+		}
+		status := ph.RunPreFilterExtensionRemovePod(ctx, state, pod, rp, nodeInfo)
+		if !status.IsSuccess() {
+			return status.AsError()
+		}
+		return nil
+	}
+	addPod := func(ap *v1.Pod) error {
+		nodeInfo.AddPod(ap)
+		status := ph.RunPreFilterExtensionAddPod(ctx, state, pod, ap, nodeInfo)
+		if !status.IsSuccess() {
+			return status.AsError()
+		}
+		return nil
+	}
+
+	podPriority := podutil.GetPodPriority(pod)
+	preemptorElasticQuotaInfo, preemptorWithElasticQuota := snapshotState.elasticQuotas[pod.Namespace]
+
+	// Check if there is elastic quota in the preemptor's namespace.
+	if preemptorWithElasticQuota {
+		moreThanMinWithPreemptor = preemptorElasticQuotaInfo.moreThanMinWithPod(preFilterState.Resource)
+	}
+
+	for _, p := range nodeInfo.Pods {
+		pElasticQuotaInfo, pWithElasticQuota := snapshotState.elasticQuotas[p.Pod.Namespace]
+		// Preemption only happens between in the same type of namespace: with or without.
+		if preemptorWithElasticQuota == pWithElasticQuota {
+			// If there is elastic quota in the preemptor's namespace.
+			if preemptorWithElasticQuota {
+				if moreThanMinWithPreemptor {
+					// If Preemptor.Request + Quota.allocated > Quota.min:
+					// It means that its guaranteed isn't borrowed by other
+					// quotas. So that we will select the pods belongs to the
+					// same quota(namespace) with the lower priority than the
+					// preemptor’s priority as potential victims in a node.
+					if p.Pod.Namespace == pod.Namespace && podutil.GetPodPriority(p.Pod) < podPriority {
+						potentialVictims = append(potentialVictims, p.Pod)
+						if err := removePod(p.Pod); err != nil {
+							return nil, 0, false
+						}
+					}
+				} else {
+					// If Preemptor.Request + Quota.allocated <= Quota.min: It
+					// means that its min or guaranteed resource is used or
+					// `borrowed` by other Quota. Potential victims in a node
+					// will be chosen from Quotas that allocates more resources
+					// than its min, i.e., borrowing resources from other
+					// Quotas.
+					if pElasticQuotaInfo.overUsed() &&
+						!pElasticQuotaInfo.lessThanMinWithoutPod(computePodResourceRequest(pod).Resource) {
```

Comment:
I have a little concern about this condition. Assume, in namespace A, we have two pods. The total used resources are larger than the Quota.min. But if we remove any of the pods, the total used resources would be smaller than the Quota.min. Then two pods would not be a preemption candidate. So, by using such a condition, it is possible that we can not find any candidate. In that case, we will break the resource guarantee.
I think that we can probably still add this kind of pod into the candidate list, but it would have less probability to be preempted. Thoughts ?

Reply:
Introduce the background, I add the check for `lessThanMinWithoutPod` for this condition:
1. we have two quotas: quota1(cpu min 3) quota2(cpu min 3) capacity is 6
2. we submit two pods(pod1A/pod1B -- cpu request 2 for every pod) for quota1, the two pods will be scheduled successfully. At this time, (quota1 used is 4, min is 3)
3. we submit two pods(pod2A/pod2B -- cpu request 2  for every pod) for quota2, pod2A is scheduled successfully. At this time,  (quota1 used is 4, min is 3) and (quota1 used is 2, min is 3). In this case, if we choose to add pod1A or pod1B as a preemption candidate, the pod1A is preempted. At this time,  (quota1 used is 2, min is 3) and (quota1 used is 4, min is 3).
4. Then pod1B is recreated by controller, then pod1B will preempt pod2A/pod2B. In this way, preempt will continue to appear between quota1 and quota2.

For this case, I won't choose the pod as  a preemption candidate if the total used resources would be smaller than the Quota.min after the pod is removed.

WDYT？Looking forward to your reply

Reply:
Ditto  
pElasticQuotaInfo is for other pods.
preemptorElasticQuotaInfo is for preemptor.

Reply:
@denkensk 
Thank you for the explanation. And yes, in the case you mentioned, the scheduler does not need to do preemption. But let us discuss the following cases. Assume, we have two quotas: quota1(CPU min 3) quota2(CPU min 3) capacity is 6.

Case1: quota1 used is 4, quota2 used is 2. When quota2 requests 2 CPU. The scheduler does not do preemption even quota1 used more resources. Because quota2 asked for more resources than the minimum resource. Also, we need to output the status/logs clearly so that the users can understand why this 2 CPU request can not be satisfied.

Case2: quota1 used is 4 (two pods, 2cpu for every pod), quota2 used is 2. When quota2 requests 1 CPU, the scheduler needs to do the preemption and delete one pod from quota1. Because we must guarantee the min resource for quota2. After that, the controller would re-request 2 CPUs for quota1, but the scheduler can not schedule it and does not do the preemption.

I understand that the preemption strategy does not sound very user-friendly. But I think that the CS should provide:
minimum resource == guaranteed resource which should/has to be satisfied at any time/ anywhere.  
max resource == best-effort resource which can not be guaranteed.

Also, we need to document the min/max concepts more clearly.
What do you think?

Reply:
Thanks for your reply.

> Case1: quota1 used is 4, quota2 used is 2. When quota2 requests 2 CPU. The scheduler does not do preemption even quota1 used more resources. Because quota2 asked for more resources than the minimum resource. Also, we need to output the status/logs clearly so that the users can understand why this 2 CPU request can not be satisfied.

Case2: quota1 used is 4 (two pods, 2cpu for every pod), quota2 used is 2. When quota2 requests 1 CPU, the scheduler needs to do the preemption and delete one pod from quota1. Because we must guarantee the min resource for quota2. After that, the controller would re-request 2 CPUs for quota1, but the scheduler can not schedule it and does not do the preemption.

Thank you for your detailed explanation. You are right. I delete the check for if we remove any of the pods, the total used resources would be smaller than the Quota.min. And also add them into the candidate list.

> Also, we need to document the min/max concepts more clearly.

Done, I add the  description about min/max concepts in README.